### PR TITLE
Feature 49 - undo/redo show action list with toolbutton

### PR DESCRIPTION
**Pull request by pboettch**
Fixes https://github.com/SciQLop/tscat_gui/issues/49

### CODE REVIEW

File: tscat_gui/__init__.py [hunk 1]
```diff
@@ -51,8 +51,180 @@ def __init__(self, parent=None):
...
+                    else:
+                        enable_move_to_trash |= True
+
+                self.restore_from_trash_action.setEnabled(enable_restore)
+                self.move_to_trash_action.setEnabled(enable_move_to_trash)
+                self.delete_action.setEnabled(True)
+                self.export_action.setEnabled(True)
+
+    def __current_event_changed(self, _: QtCore.QModelIndex, __: QtCore.QModelIndex):
+        if self.programmatic_select:
```
Comment (jeandet) [inline tscat_gui/__init__.py:117]: an if not seems way better

File: tscat_gui/__init__.py [hunk 2]
```diff
@@ -51,8 +51,180 @@ def __init__(self, parent=None):
...
+
+    def __catalogue_selection_changed(self, _: QtCore.QItemSelection, __: QtCore.QItemSelection) -> None:
+        if self.programmatic_select:
+            return
+
+        uuids = [index.data(UUIDRole) for index in self.catalogues_view.selectedIndexes()]
+        self.state.updated('active_select', tscat._Catalogue, uuids)
+
+    def __undo_redo_index_changed(self, index: int) -> None:
+        def __create_action_list_menu(index_range: range,
```
Comment (jeandet) [inline tscat_gui/__init__.py:131]: this could be a module free function

File: tscat_gui/__init__.py [hunk 3]
```diff
@@ -51,8 +51,180 @@ def __init__(self, parent=None):
...
+                                           "Catalogue import",
+                                           f"The selected file could not be imported: '{e}'.")
+
+    def __export_to_file(self):
+        filename, filetype = QtWidgets.QFileDialog.getSaveFileName(
+            self,
+            "Specify the filename for exporting the selected catalogues",
+            str(Path.home()),
+            "JSON Document (*.json)")
+        if filename == '':
```
Comment (jeandet) [inline tscat_gui/__init__.py:205]: this could be replaced by an if not too

**Review by jeandet — CHANGES_REQUESTED** [review]
This is better, one more function could be moved and early returns should be removed.

File: tscat_gui/__init__.py (hunk 1 above)
Comment (pboettch) [reply tscat_gui/__init__.py:117]: I love early returns. Much better than a whole tree of `if <true-conditions>` (not the case here I admit) and indented code.

What's the problem?

File: tscat_gui/__init__.py (hunk 2 above)
Comment (pboettch) [reply tscat_gui/__init__.py:131]: You mean, I could put this function somewhere else? It captures and uses `self`.

Comment (pboettch) [reply tscat_gui/__init__.py:131]: It captures and uses  `self`. What do you mean by module-free?

File: tscat_gui/__init__.py (hunk 1 above)
Comment (jeandet) [reply tscat_gui/__init__.py:117]: It both doesn't give any performance benefit nor improves readability. The important piece of code is not the "return None" but the one actually doing something here (L120 and 121). So I don't see why you want to make it more complicated by making implicit the fact that it should only be evaluated when self.programmatic_select is False.

File: tscat_gui/__init__.py (hunk 2 above)
Comment (jeandet) [reply tscat_gui/__init__.py:131]: free apply to function, if there is no way to make a function then a method it should be.

File: tscat_gui/__init__.py (hunk 1 above)
Comment (pboettch) [reply tscat_gui/__init__.py:117]: So it's a question of taste and habits. I'll change it to the style you prefer.

**Review by jeandet — APPROVED** [review]
Looks ok to me

**Pull request merged**